x-pack/plugins/spaces/server/new_platform/spaces_service/spaces_service.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { first } from 'rxjs/operators';
+import { Observable, Subscription } from 'rxjs';
+import { Legacy } from 'kibana';
+import { SavedObjectsService } from 'src/legacy/server/kbn_server';
+import {
+  Logger,
+  ElasticsearchServiceSetup,
+  HttpServiceSetup,
+  KibanaRequest,
+} from 'src/core/server';
+import { DEFAULT_SPACE_ID } from '../../../common/constants';
+import { SecurityPlugin } from '../../../../security';
+import { SpacesClient } from '../../lib/spaces_client';
+import { getSpaceIdFromPath } from '../../lib/spaces_url_parser';
+import { SpacesConfigType } from '../config';
+
+type RequestFacade = KibanaRequest | Legacy.Request;
+
+export interface SpacesServiceSetup {
+  scopedClient(request: RequestFacade): SpacesClient;
+
+  getSpaceId(request: RequestFacade): string;
+
+  isInDefaultSpace(request: RequestFacade): boolean;
+}
+
+interface CacheEntry {
+  spaceId: string;
+  isInDefaultSpace: boolean;
+}
+
+interface SpacesServiceDeps {
+  http: HttpServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+  savedObjects: SavedObjectsService;
+  getSecurity: () => SecurityPlugin | undefined;
+  config$: Observable<SpacesConfigType>;
+  spacesAuditLogger: any;
+}
+
+export class SpacesService {
+  private configSubscription$?: Subscription;
+
+  private readonly contextCache: WeakMap<any, CacheEntry> = new WeakMap();
+
+  constructor(private readonly log: Logger, private readonly serverBasePath: string) {
+    this.contextCache = new WeakMap();
+  }
+
+  public async setup({
+    http,
+    elasticsearch,
+    savedObjects,
+    getSecurity,
+    config$,
+    spacesAuditLogger,
+  }: SpacesServiceDeps): Promise<SpacesServiceSetup> {
+    let config: SpacesConfigType = await config$.pipe(first()).toPromise();
```

Comment:
I modeled this off of similar code in core, but it admittedly doesn't then go and subscribe to config changes. Do you have any recommendations? https://github.com/elastic/kibana/blob/461a6c0f9345110de23beee3381eac8f4376da03/src/core/server/http/http_service.ts#L76

Reply:
@elastic/kibana-platform do you all have any recommendations here? The `config$` is actually a `BehaviorSubject` so what we're doing here technically works, but since the type is just an `Observable`, it doesn't seem like something we should be relying upon.

Reply:
`elaticsearch service` exposes clients as Observables as well https://github.com/restrry/kibana/blob/f1468da0d95d44d80645cfc21ed44769e2d99680/src/core/server/elasticsearch/elasticsearch_service.ts#L102
later client can be bound to a specific request https://github.com/restrry/kibana/blob/f1468da0d95d44d80645cfc21ed44769e2d99680/src/core/server/elasticsearch/cluster_client.ts#L155

Reply:
I guess the same question applies to the elasticsearch service then :) When we want to break out of the RxJs paradigm and write code imperatively within the context of http request handlers, do you have a recommended approach to doing so?

Reply:
seems that you **can** use `first` or `take(1)` to get the current value, because it assumes `first` at the execution time.
example:
https://codesandbox.io/s/f0oyn?fontsize=14

Reply:
The part I find confusing, is if we were to use a `from("first value", "second value", "third value")` instead of the BehaviorSubject, we always see the very first value per https://codesandbox.io/embed/relaxed-keller-kqvgc

Reply:
This is the "behaviour" of a BehaviourSubject (vs an ordinary subject). 

> BehaviorSubjects are useful for representing "values over time". For instance, an event stream of birthdays is a Subject, but the stream of a person's age would be a BehaviorSubject.

> One of the variants of Subjects is the BehaviorSubject, which has a notion of "the current value". It stores the latest value emitted to its consumers, and whenever a new Observer subscribes, it will immediately receive the "current value" from the BehaviorSubject.

Consumers of `config$` don't care about previous stale configs when they subscribe, they just want the most up-to-date config, and that's the first value they'll get.

http://reactivex.io/rxjs/manual/overview.html#behaviorsubject

Reply:
I think I misunderstood the question.

We expose BehaviourSubjects as Observables because exposing the BehaviourSubject would also allow the consumer to publish values like `config$.next('lol')`.

But I agree with you that this doesn't make for a great API, you need to know the internal implementation details to be able to consume it properly.

The only thing I can think of is a naming convention like `latestConfig$` and then we should specify in the docs that a given observable is e.g. a shared replay subject / behaviour subject.

Reply:
I created an issue to discuss alternative ways to make this more clear and track implementation https://github.com/elastic/kibana/issues/38397

Reply:
Thanks @skaapgif !

Reply:
Resolved the usage here in https://github.com/elastic/kibana/pull/35429/commits/3a5cc82cdcb90725f6a1019539e430340c6bd471